### PR TITLE
Add reference name to queries

### LINKED ISSUE #180: Add reference_name to Queries
## Is your feature request related to a problem? Please describe

When adding a query to the Queries section the use should set a reference name for the query to use inside any scheduled templates

## Acceptance Criteria

- [ ] reference name must be unique
- [ ] reference name is required
- [ ] reference name should be a valid identifier for tera

## Describe alternatives you've considered

Set a reference name inside each scheduled notification

## Additional context

Is there a better name than `reference_name`?

### PR DESCRIPTION
**Pull request by jmbrunskill**
Closes #180 

# 👩🏻‍💻 What does this PR do?
 
Allows the user to set a reference name for a query, to be used in scheduled notifications (needed for #154 )

image

image

# 🧪 How has/should this change been tested?

Should be possible to try creating, and editing queries with reference names.
Reference name should be unique so would be good to test that.

### CODE REVIEW

File: backend/repository/migrations/2023-10-09-042324_add_reference_name/down.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+-- This file should undo anything in `up.sql`
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/repository/migrations/2023-10-09-042324_add_reference_name/down.sql:1]: Yeah, I have a similar opinion, not planning on supporting down migrations it doesn't make sense to put the effort in for our situation (i believe), if you did roll back would you want to loose data from what column for example?

BTW: This is the default text that diesel generates, I just didn't change it.

File: backend/graphql/notification_query/src/mutations/create.rs [hunk 2]
```diff
@@ -41,7 +42,17 @@ pub fn create_notification_query(
...
+            id,
+            name,
+            reference_name,
+        }: CreateNotificationQueryInput,
+    ) -> Self {
+        CreateNotificationQuery {
+            id,
+            name,
+            reference_name,
+        }
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/graphql/notification_query/src/mutations/create.rs:56]: True, that would easier, I was just copying existing patterns. I remember this being super annoying at the start, feels like it could be a derive macro or something in many cases!!

File: backend/repository/migrations/2023-10-09-042324_add_reference_name/up.sql [hunk 3]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE notification_query ADD COLUMN reference_name TEXT NOT NULL default id;
+UPDATE notification_query SET reference_name = id;
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/repository/migrations/2023-10-09-042324_add_reference_name/up.sql:2]: Yes it seemed to be, but I'm also not entirely sure why, it appears the default didn't run immediately or something?
It's a bit of a edge case anyway mostly it's only dev/test machines that would even update this.

File: backend/repository/src/db_diesel/notification_query.rs [hunk 4]
```diff
@@ -100,17 +101,29 @@ fn create_filtered_query(filter: Option<NotificationQueryFilter>) -> BoxedQuery
...
 
         if let Some(search) = search {
             let search_term = format!("%{}%", search);
             query = query.filter(
                 notification_query_dsl::name
                     .like(search_term.clone())
-                    .or(notification_query_dsl::description.like(search_term)),
+                    .or(notification_query_dsl::description.like(search_term.clone()))
+                    .or(notification_query_dsl::reference_name.like(search_term.clone()))
+                    .or(notification_query_dsl::query.like(search_term.clone())),
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/repository/src/db_diesel/notification_query.rs:126]: Yes, I think you are right, good catch. I'm not sure when or if we'd use a query like you're describing but it I think it will generate incorrect queries.

**Review by clemens-msupply — APPROVED** [review]
Works great, thanks!

One thing user might get confused by is why the reference name is needed or what it is used for. Maybe a short description what it the label?

File: backend/repository/src/db_diesel/notification_query.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/repository/src/db_diesel/notification_query.rs:126]: I take it back, I think it's ok because it's inside a new `.filter` call?

```
SELECT * FROM `notification_query` WHERE `notification_query`.`id` = ? AND (((`notification_query`.`name` LIKE ? OR `notification_query`.`description` LIKE ?) OR `notification_query`.`reference_name` LIKE ?) OR `notification_query`.`query` LIKE ?) ORDER BY `notification_query`.`id` ASC LIMIT ? OFFSET ? -- binds: ["98533e08-99bb-4b18-a045-db67e1852d73", "%%", "%%", "%%", "%%", 100, 0]
```

Seems like `.filter` is effectively `AND`

> Adds to the WHERE clause of a query. If there is already a WHERE clause, the result will be old AND new.

File: backend/service/src/notification_query/mod.rs [hunk 5]
```diff
@@ -106,6 +110,15 @@ impl PartialEq for ModifyNotificationQueryError {
...
                 ModifyNotificationQueryError::InvalidNotificationQueryName,
             ) => true,
+            (
+                ModifyNotificationQueryError::InternalError(self_err),
+                ModifyNotificationQueryError::InternalError(other_err),
+            ) => self_err == other_err,
+            (
+                ModifyNotificationQueryError::BadUserInput(self_err),
+                ModifyNotificationQueryError::BadUserInput(other_err),
+            ) => self_err == other_err,
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/service/src/notification_query/mod.rs]: It works fine!!!
I think it had a problem with this one time, and ended up with this that's being copy and pasted everywhere.
Will check for others that can be removed!

File: backend/service/src/notification_query/validate.rs [hunk 6]
```diff
@@ -46,6 +46,26 @@ pub fn check_notification_query_name_is_unique(
...
 
+pub fn check_notification_query_reference_name_is_unique(
+    id: &str,
+    reference_name: Option<String>,
+    connection: &StorageConnection,
+) -> Result<bool, RepositoryError> {
+    match reference_name {
+        None => Ok(true),
+        Some(reference_name) => {
+            let notification_queries = NotificationQueryRepository::new(connection)
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply backend/service/src/notification_query/validate.rs]: Thanks, that's nicer!

**Comment by jmbrunskill** [comment]
> Works great, thanks!
> 
> One thing user might get confused by is why the reference name is needed or what it is used for. Maybe a short description what it the label?

Yes a very good idea. I had meant to do that!

File: backend/repository/src/db_diesel/notification_query.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply backend/repository/src/db_diesel/notification_query.rs:126]: oh yes that should work then, thanks

File: backend/graphql/notification_query/src/mutations/create.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply backend/graphql/notification_query/src/mutations/create.rs:56]: Yes think a macro would be possible (date, datetimes may cause some problems though). The mapping is a bit annoying but with destructuring the input it's IMHO not too bad.
